Enhance/two speedometers

braem opened the pull request:
Closes #558 

<ul>
<li>Added a horizontal speedometer</li>
<li>Moved last jump velocity display into it's own class</li>
<li>Created a superclass for speedometer classes to inherit</li>
<ul>
<li>Couldn't move much functionality into the superclass as the speedometer subclasses have to be the ones inheriting from hudelement and hudnumericdisplay</li>
<li>Each speedometer uses the same units and colorize setting.</li>
</ul>
<li>Added U/S unit label and an option to remove unit labels</li>
<li>Dynamically set hudelements so that any combination of active speedometers displays without large gaps & indents. There is at most 1 unit label shown at a time, above the top-most speedometer.</li>
<ul>
<li>When both the absolute and horizontal speedometer are active, the horizontal one always displays at the top. </li>
</ul>
<li>Updated settings in the HUD settings tab to reflect these changes.</li>
</ul>

Ex:
![image](https://user-images.githubusercontent.com/9014762/76282428-d8a9c100-6254-11ea-97ed-6d87ac87ed01.png)

Settings Page:
![image](https://user-images.githubusercontent.com/9014762/76281250-b2365680-6251-11ea-81a2-8bae1b0608a0.png)

### Checklist
- [x] If there is a localization token change, I have updated the `momentum_english_ref.res` file with the changes, ran `tokenizer.py` to generate an up-to-date localization file, and have committed both the `.res` file changes and the new localization `.txt` file
- [x] If I introduced new h/cpp files, I have added them to the appropriate project's VPC file (`server_momentum.vpc` / `client_momentum.vpc` / etc)
- [x] My commits are relatively small and scoped to the best of my ability
- [x] My branch has a clear history of changes that can be easy to follow when being reviewed commit-by-commit
- [x] My branch is functionally complete; the only changes to be done will be those potentially requested in code review

Gocnak commented:
Alright, this is a really good first attempt and honestly the speedometer code is probably the oldest HUD element we have so I don't blame you for tackling such an ugly mess that was already there. I'm seeing a lot of code duplication though, and it seems pretty confusing to make all of the speedometer-related elements each their own element.

Since this was originally for 0.8.5 my original plan was going to look into cleaning up this class altogether; the HudNumbers class is nice but given how much we actually want to do compared to the original use of HudNumbers (being HL2's "Health" and "Armor" and "Ammo" HUD elements), I think it's wise to depart from it and instead start to use EditablePanel (and its `.res` files), and Labels.

My vision for this is to have a stack of labels:
```
=====================
|   Units Label      |
|   Absolute         |
|   Horizontal       |
|   Last Jump        |
|   Stage Enter/Exit |
=====================
```

That automatically "pin" to each other based on what's toggled to be visible. So if `Units Label` isn't visible, `Absolute` would be up top with everything below it; if `Horizontal` isn't toggled on then `Last Jump` and `Stage Enter/Exit` would shift upwards to take its place, etc.

The speedometer element would stay as a CHudElement, but rather inherit from `EditablePanel` instead of `HudNumbers`, and each of the actual displays would just be labels as children of it. The individual labels would have their settings be defined in a `.res` file that is loaded in the constructor. See [the timer](https://github.com/momentum-mod/game/blob/develop/mp/src/game/client/momentum/ui/HUD/hud_timer.cpp) if you want a good example of this.

The custom layout will most likely need to be done in a `PerformLayout` override though, and will be a fun one to handle. In fact we may benefit from having it be under a `mom_hud_speedometer_autolayout` toggle var, so that players can still completely customize it to their liking.

braem commented:
Originally I planned on shoving most of the code into a speedometer superclass, but I found that the hud elements didn't work unless they were directly inherited. I avoided changing the higher up classes as I wasn't sure how they would effect other things on the HUD. I will try all of these recommendations.

As for custom layout, Zike mentioned that jumpers may want the speedometer in another location: https://i.imgur.com/rBq7Bjs.png . Also issue #444 mentions surfers would like the option to have the absolute speedometer in the center of the screen. So definitely a `mom_hud_speedometer_autolayout` should be there to allow this.

I'm not sure on the time frame for this as I work an ITS job in health care and things are picking up due to the coronavirus. I will work on this as soon as I can though :)

braem commented:
Switched implementation to use `EditablePanel` instead of `HudNumbers`, and cleaned up duplicate code by moving everything into one hud element and using labels as requested. The absolute speedometer also displays above the horizontal one.
Created a `speedometer.res` to store the label resources.

The custom layout should be implemented when addressing #444

braem commented:
Rebased onto develop. 
Fixed some more issues:
- Found `hud_comparisons` and `hud_mapfinished` were referencing the removed `mom_hud_speedometer_hvel` cvar. Changed them to use a new cvar: `mom_hud_velocity_type`
- Changed fonts & formatting to match the old speedometer code
- Fixed last jump vel not fading out
- Fixed formatting when running `hud_reloadscheme`
- Fixed stage enter velocity not displaying in the center when no comparisons were loaded

The biggest issue I find is that I'm not sure where to put this on the hud. It does not fit between the checkpoint and sticky charge meter when everything is active. The sticky charge and sticky count hud elements may need to be moved up to accommodate this.
Not sure if the decision is mine to make. Should be an easy change regardless as all functionality is there now.

Gocnak commented:
> The sticky charge and sticky count hud elements may need to be moved up to accommodate this.

Agreed. Do that for now. I would suggest doing `pin_to_sibling` as well for the elements, but that can come at a later date when we get some UX eyes on the entire HUD as is.

braem commented:
> 
> I would suggest doing `pin_to_sibling` as well for the elements, but that can come at a later date when we get some UX eyes on the entire HUD as is.

`pin_to_sibling` breaks the dynamically set y positions of the speedometers. 
Could dynamically set y positions of the other elements based on how many speedometers are present (snap them downward when one speedometer is shut off & vice versa). Seems odd to do this in the `hud_speedometer.cpp` class file though.. would that be alright?

For now I just hard coded the positions of these elements as they were before.

braem commented:
Deleted the commit with the hard coded positions.
Implemented `pin_to_sibling` by pinning the sticky count to the sticky charge meter and the charge meter to the speedometers. This way both the sticky count and sticky charge meter move downward when a speedometer is disabled & vice versa.
I reduced the spacing between these elements to have them fit well.

This is how it looks with all speedos active:
![AllElements](https://user-images.githubusercontent.com/9014762/77809091-71a94c00-704b-11ea-9772-48beaafca0ee.png)

And this is when unit labels, last jump vel, and enter speed are disabled:
![onlySpeedos](https://user-images.githubusercontent.com/9014762/77809114-88e83980-704b-11ea-974e-ca9ceb60d86a.png)
I included this screenie as I believe this is what jumpers would prefer. Last jump vel and showing enter speed are not that useful for sticky jump (in my opinion). Enter speed could be useful.
Also this solves something Zike pointed out in passing on his stream: the sticky charge meter is too close to the center of the screen.

I would suggest adding the show enter speed as a setting on the hud settings page as well. Would this be alright?

Gocnak commented:
> I would suggest adding the show enter speed as a setting on the hud settings page as well. Would this be alright?

Huh was it not already? Yeah absolutely, add it to the HUD settings page.

braem commented:
> > I would suggest adding the show enter speed as a setting on the hud settings page as well. Would this be alright?
> 
> Huh was it not already? Yeah absolutely, add it to the HUD settings page.

Yeah I remember showing Zike the convar when he was saying there was no way to disable it. I added it to the HUD settings. Added some tooltips to clarify functionality & logic to disable/enable setting fields correctly as well.

Also fixed a bug where the sticky charge meter & count didn't snap down when disabling multiple speedos at a time in the settings panel (before I was just testing individual cvars).

braem commented on mp/game/momentum/resource/ui/Speedometer.res:
```diff
@@ -0,0 +1,172 @@
+        "auto_tall_tocontents" "1"
+		"auto_wide_tocontents" "1"
+	}
+}
```
Changed to use only spaces. A lot of the fields are copied over from other resource files so I had a look at them as well and there are a lot of tabs in those.

braem commented on mp/game/momentum/scripts/HudLayout.res:
```diff
@@ -393,10 +377,14 @@
     {
         "fieldName"		"HudStickyCharge"
+        "pin_to_sibling" "HudSpeedMeter"
+        "pin_to_sibling_corner" "4"
```
Changed to pin to correct corners. 4 for top center and 6 for bottom center. Had to adjust wide/tall of other elements for this to work well, as they were much wider/taller than they need to be.

braem commented on mp/game/momentum/scripts/HudLayout.res:
```diff
@@ -1,33 +1,17 @@
+	{
+		"fieldName"		"HudSpeedMeter"
+        // See resource/ui/Speedometer.res for changing settings
+	}
```
Fixed. See other comment relating to tabs though.

Gocnak replied:
This doesn't seem fixed.

braem replied:
woops, only changed the fieldname. Done now.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -26,93 +31,161 @@ using namespace vgui;
+
     void Paint() OVERRIDE;
-    void PaintNumbers(HFont font, int xpos, int ypos, int value, bool atLeast2Digits) OVERRIDE;
+    //void PaintNumbers(HFont font, int xpos, int ypos, int value, bool atLeast2Digits) OVERRIDE;
```
My bad, I forgot to remove before commiting.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -26,93 +31,161 @@ using namespace vgui;
+    m_defaultYPos = this->GetYPos();
+}
+
+void CHudSpeedMeter::inactivateLabel(Label *label)
```
Fixed this an `DeactiveLabel`

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -294,132 +424,151 @@ void CHudSpeedMeter::OnThink()
+    if (!mom_hud_speedometer_lastjumpvel.GetBool()) yIndent += m_defaultLastJumpVelLabelHeight;
+    if (!mom_hud_speedometer_unit_labels.GetBool()) yIndent += m_defaultUnitsLabelHeight;
+    if (!mom_hud_speedometer_showenterspeed.GetBool()) yIndent += m_defaultStageEnterExitLabelHeight;
+    this->SetPos(this->GetXPos(), m_defaultYPos + yIndent);
```
Apologies, I do this a lot in my own projects. Didn't know this was enforced.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -139,19 +212,27 @@ void CHudSpeedMeter::FireGameEvent(IGameEvent *pEvent)
+bool CHudSpeedMeter::ShouldDraw() 
+{ 
+    return CHudElement::ShouldDraw();
 }
```
Was using this before, but the override became useless as I let the `Paint` function display what needs to be displayed.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -160,131 +241,180 @@ void CHudSpeedMeter::OnThink()
 
         // only called if we need to update color
-        if (mom_hud_speedometer_colorize.GetInt())
+        if (ConVarRef("mom_hud_speedometer_colorize").GetInt())
```
Holdover from when the convars were defined in separate files. Reverted these back.

braem commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -139,19 +212,27 @@ void CHudSpeedMeter::FireGameEvent(IGameEvent *pEvent)
-    BaseClass::ApplySchemeSettings(pScheme);
-
-    m_SecondaryColor = GetSchemeColor("SecondaryValueColor", pScheme);
+    Panel::ApplySchemeSettings(pScheme);
```
Switched it back, shouldn't have been changed. I believe I was looking at [hud_comparisons.cpp](https://github.com/momentum-mod/game/blob/develop/mp/src/game/client/momentum/ui/HUD/hud_comparisons.cpp) at the time, but that was inheriting `Panel` not `EditablePanel`. Not sure why that isn't using `BaseClass` though.

braem commented:
Implemented the requested changes.

Couple of things of note:
- Found that a lot of the .res files have tabs intermixed with spaces
- Had to modify the wide/tall fields of sticky count and charge meter to properly pin elements. Looks good now with them having `xpos=0` and `ypos=0`.
- Remembered that, when I looked at [hud_comparisons.cpp](https://github.com/momentum-mod/game/blob/develop/mp/src/game/client/momentum/ui/HUD/hud_comparisons.cpp), `Panel::` is used instead of `BaseClass::`, though it wouldn't change anything.

As for the refactoring, yes this class needs it badly. I understand what most of the `OnThink` function is doing so I could tackle that, if you aren't putting someone with more experience/someone on the team on it. Understandable if so as this PR has been a mess.

braem commented:
Squashed the earlier commits using a `fixup` on them in interactive rebase. These commits were from before moving to using an `EditablePanel` inheritance. Not very useful as at that time I was splitting the speedometers across multiple source files.

Gocnak commented on mp/game/momentum/resource/ui/SettingsPanel_HudSettings.res:
```diff
@@ -121,10 +151,40 @@
+        "use_proportional_insets"		"0"
+        "Default"		"0"
+        "cvar_name"		"mom_hud_speedometer_unit_labels"
+        "cvar_value"		"1"
```
Since this is a new thing you're adding I would remove the `cvar_value` from just this section, and worry about the rest of the ones in your #639 PR.

braem replied:
Removed all the cvars in this file using find/replace.. woops. Shouldn't effect #639 very much though.

Gocnak commented on mp/game/momentum/resource/ui/SettingsPanel_HudSettings.res:
```diff
@@ -121,10 +151,40 @@
-        "cvar_name"		"mom_hud_speedometer_hvel"
-        "cvar_value"		"0"
+        "cvar_name"		"mom_hud_speedometer_showenterspeed"
+        "cvar_value"		"1"
```
Same here, I would remove cvar_value since you're already touching this in this PR, but no others.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -246,180 +329,256 @@ void CHudSpeedMeter::OnThink()
-                else
+                else //color based on relation to max vel
                 {
                     const float maxvel = ConVarRef("sv_maxvelocity").GetFloat();
```
The main thing to note about `ConVarRef` is that it's not *too* bad if it's setup once (constructor or as a static), but *every frame* can cause a significant performance problem, as it constantly looks up the convar using strings and linear looping...

It's an artifact of our old code, but if you're in the area, definitely try to either:
1. Directly use the reference, if possible. This means `extern ConVar <convarname>;` or in this instance, including the header where it already is extern'd.
2. Use `static ConVarRef <convarname>("<convarname>");` if 1 is not possible.

braem replied:
Oh never thought of that. 
Speedometer was already including `movevars_shared.h` but was using `ConVarRef` for `sv_maxvelocity`, so I removed that. Also moved the reference to `mom_replay_timescale` from a local `ConVarRef` to the constructor. Did this for the new `mom_hud_velocity_type` in the map finished panel as well. The comparisons panel was already doing this so no changes there.

Gocnak commented:
> so I could tackle that, if you aren't putting someone with more experience/someone on the team on it.

Yeah the main thing is that we're going to want to completely overhaul the speedometer as part of #638 , which can be split up into multiple parts honestly, with part of it being refactoring code we have, and then the second part being the customization. Feel free to tackle the first part of it.

braem commented on mp/game/momentum/resource/momentum_english_ref.res:
```diff
@@ -274,17 +274,21 @@
-        "MOM_Settings_Speedometer_Units_UPS" "Units per second"
-        "MOM_Settings_Speedometer_Units_KPH" "Kilometers per hour"
-        "MOM_Settings_Speedometer_Units_MPH" "Miles per hour"
+        "MOM_Settings_Speedometer_Units_UPS" "Units per second (U/S)"
```
Switched the localization, units label & cvar descriptions to use UPS instead of U/S.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -26,93 +31,160 @@ using namespace vgui;
-static MAKE_TOGGLE_CONVAR(
-    mom_hud_speedometer_showenterspeed, "1", FLAG_HUD_CVAR,
-    "Toggles showing the stage/checkpoint enter speed (and comparison, if existent). 0 = OFF, 1 = ON\n");
+static MAKE_CONVAR(mom_hud_velocity_type, "0", FLAG_HUD_CVAR | FCVAR_CLIENTCMD_CAN_EXECUTE,
+                   "Toggles the velocity type used in comparisons and map finished dialog. 0 = ABSOLUTE, 1 = HORIZONTAL",
+                   0, 1);
```
```suggestion
static MAKE_TOGGLE_CONVAR(mom_hud_velocity_type, "0", FLAG_HUD_CVAR | FCVAR_CLIENTCMD_CAN_EXECUTE,
                   "Toggles the velocity type used in comparisons and map finished dialog. 0 = ABSOLUTE, 1 = HORIZONTAL");
```

braem replied:
Reasoning for this was to make room for the possibility of different speed types like vertical only, but that is unlikely to happen anyway. Made the requested change.

Gocnak commented on mp/src/game/client/momentum/ui/HUD/hud_speedometer.cpp:
```diff
@@ -26,93 +31,159 @@ using namespace vgui;
+    
+    LoadControlSettings("resource/ui/Speedometer.res");
+
+    m_defaultAbsSpeedoLabelHeight = m_pAbsSpeedoLabel->GetTall();
+    m_defaultHorizSpeedoLabelHeight = m_pHorizSpeedoLabel->GetTall();
+    m_defaultLastJumpVelLabelHeight = m_pLastJumpVelLabel->GetTall();
+    m_defaultUnitsLabelHeight = m_pUnitsLabel->GetTall();
+    m_defaultStageEnterExitLabelHeight = m_pStageEnterExitLabel->GetTall();
+
+    m_pStageEnterExitComparisonLabel->SetFont(m_pStageEnterExitLabel->GetFont()); //need to have same font
+
+    m_defaultYPos = GetYPos();
```
Duplicate code found in `ApplySchemeSettings`, can this be removed?

Gocnak replied:
Or if you really need to initialize them, pull this and the code in `ApplySchemeSettings` out to a private member function.

braem replied:
Ah, didn't realize `ApplySchemeSettings` was called after constructor. I was under the impression it was called when using `hud_reloadscheme`.
Removed this useless code.
